[Profiling] Deep-linking

dgieselaar opened the pull request:
Adds deeplinking for Profiling app.

<img width="1472" alt="CleanShot 2022-07-07 at 13 37 08@2x" src="https://user-images.githubusercontent.com/352732/177764547-3be28c15-e867-4cd1-a5f2-fa7086275728.png">

<img width="1473" alt="CleanShot 2022-07-07 at 13 37 29@2x" src="https://user-images.githubusercontent.com/352732/177764584-f4ac9487-7c07-4d0d-9a7e-6d32defd47b7.png">

rockdaboot reviewed (approved):
:+1: Looks so much better !

We have to talk soon about the differential flamegraph and how we want to implement it in the frontend  and backend. Currently, prodfiler just does 2 parallel flamegraph queries and the client juggles with a lot of data. Maybe we can improve this.

dgieselaar commented on x-pack/plugins/profiling/common/runtime_types/range_rt.ts:
```diff
@@ -0,0 +1,12 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import * as t from 'io-ts';
```
`io-ts` is a runtime validation library that is also type-safe. We use it in `@kbn/typed-react-router-config`, which is the library that we use for type-safe client-side routing.

dgieselaar commented on x-pack/plugins/profiling/common/types.ts:
```diff
@@ -20,8 +20,4 @@ export interface IMyStrategyResponse extends IEsSearchResponse {
 export interface TimeRange {
   start: string;
   end: string;
-  isoStart: string;
-  isoEnd: string;
-  unixStart: number;
-  unixEnd: number;
```
I've removed these, `start` and `end` are ISO 8601.

dgieselaar commented on x-pack/plugins/profiling/public/components/flame_graphs_view/index.tsx:
```diff
@@ -0,0 +1,72 @@
+  const {
+    query,
+    query: { rangeFrom, rangeTo, index, projectID, n },
+  } = useProfilingParams('/flamegraphs/*');
```
`useProfilingParams` will return typed parameters for the given route. If the component is being rendered but the given is not active, the hook will throw an error.

dgieselaar commented on x-pack/plugins/profiling/public/components/profiling_app_page_template/index.tsx:
```diff
@@ -5,59 +5,80 @@
+    path,
+    query,
+    query: { rangeFrom, rangeTo, n, projectID, index },
+  } = useProfilingParams('/*');
```
`/*` is a greedy match, it will match all routes.

dgieselaar commented on x-pack/plugins/profiling/public/routing/index.tsx:
```diff
@@ -0,0 +1,94 @@
+
+const routes = {
+  '/': {
+    element: <Outlet />,
```
`<Outlet/>` is where children of a route are rendered.

jbcrail reviewed (commented):
The following feedback is related to the overall UI/UX and the behavior when compared to the prodfiler UI. I don't have any specific feedback about the code, which looks good to me.

1. Default to stacktraces view on startup

The existing prodfiler UI currently defaults to the stacktrace view because it's used as a high-level overview for users before they drill down into specific trace details. Also, since the flamegraph query is heavier than most of the stacktrace queries, users have a shorter time-to-interaction response when the stacktrace view is the default.

2. Remember selected time range when switching between subviews

When exploring activity for a given time period, a user will often switch between stacktrace, flamegraph, top functions, and callgraph. In the existing prodfiler UI, we remember the selected time range so that can see different views of the data without needing to manually set the time range every time they switch. We would like to retain this behavior if possible. I'm not sure if using the navigation sidebar is a hindrance to this.

3. Add navigation breadcrumbs in the top bar

I only see `Kibana` at the moment in the breadcrumb section. It would be nice if we could have similar functionality like other Observability sections.

4. Reduced horizontal viewing area when using sidebar as navigation

This is more of an observation and not a request for specific changes at the moment. I just wanted to get my thoughts down for possible future design considerations. From a user's perspective, using the sidebar for navigation between views has tradeoffs. The navigation sidebar is convenient for hiding possible clutter and makes it easier to add more views by using vertical space instead of the constrained space near the top. However, it does come at the cost of discoverability and ease of access to the views (stacktrace, flamegraph, callgraph, etc) when the sidebar is collapsed.

dgieselaar commented:
@jbcrail that's great feedback, thanks.

> Default to stacktraces view on startup
> The existing prodfiler UI currently defaults to the stacktrace view because it's used as a high-level overview for users before they drill down into specific trace details. Also, since the flamegraph query is heavier than most of the stacktrace queries, users have a shorter time-to-interaction response when the stacktrace view is the default.

I've changed the default route to `/stacktraces/containers`.

> Remember selected time range when switching between subviews
> When exploring activity for a given time period, a user will often switch between stacktrace, flamegraph, top functions, and callgraph. In the existing prodfiler UI, we remember the selected time range so that can see different views of the data without needing to manually set the time range every time they switch. We would like to retain this behavior if possible. I'm not sure if using the navigation sidebar is a hindrance to this.

Thanks! Fixed.

> Add navigation breadcrumbs in the top bar
> 	I only see Kibana at the moment in the breadcrumb section. It would be nice if we could have similar functionality like other Observability sections.

Right, good point. I need to move some stuff around for that to work, will address it in a follow-up PR.

> Reduced horizontal viewing area when using sidebar as navigation
> This is more of an observation and not a request for specific changes at the moment. I just wanted to get my thoughts down for possible future design considerations. From a user's perspective, using the sidebar for navigation between views has tradeoffs. The navigation sidebar is convenient for hiding possible clutter and makes it easier to add more views by using vertical space instead of the constrained space near the top. However, it does come at the cost of discoverability and ease of access to the views (stacktrace, flamegraph, callgraph, etc) when the sidebar is collapsed.

Agree, I also discussed this with @boriskirov briefly. In the case of the Profiling links, they're at the bottom, so it's easy to miss them even when the sidebar is open. I don't think we have a good solution for that today but it's something that we should probably address.
